clippy_lints/src/attrs.rs

```diff
@@ -716,6 +756,7 @@ impl EarlyLintPass for EarlyAttributes {
         check_mismatched_target_os(cx, attr);
         check_minimal_cfg_condition(cx, attr);
         check_misused_cfg(cx, attr);
+        check_should_panic_reason(cx, attr);
```

Comment:
The lint is in the `Attributes` pass but the check is in `EarlyAttributes`.

This should be in `Attributes` instead anyway as [early lints in this file are pre-expansion lints](https://github.com/rust-lang/rust-clippy/blob/ee8a429792eb280d20c5e4f4b1549d1dd3b30335/clippy_lints/src/lib.rs#L370), which have historically had countless problems and are quasi-deprecated in rustc. The implementation should work the same as a late pass lint

Reply:
The reason I made it part of the early lint pre-expansion pass is because that makes finding the `#[should_panic]` attribute a lot easier. `#[test]` seems to expand to a `const` item with the `should_panic` attribute part of a field of the struct being initialized in the const item somewhere.
`#[should_panic]` doesn't appear in `Attribute`'s `check_attribute` fn, so I guess we'd have to look for these kinds of const items, which sounds a lot more painful than what it's currently doing. Or am I missing something?

Reply:
Oh yeah that's a good point actually, also more importantly `#[cfg(test)]` and just `#[test]` will cause it to not appear in the HIR *unless* you're running `cargo test`

So this'll unfortunately need to be a pre-expansion lint. To make sure that `#[cfg(test)}` and all work tests should be added for those

Reply:
If it can be done outside of a pre expansion pass it should be, the main pre expansion bug is it not picking up lint levels correctly - https://github.com/rust-lang/rust-clippy/issues/6610, for a lint that you might want to globally allow this is particularly bad

`cargo clippy --tests` (or `--all-targets` etc) can be used to run clippy on tests

Reply:
Right, okay, that's pretty bad. This makes it quite a bit more complicated, so I suppose we'll have to look for the expanded `#[test]` const item and extract the `should_panic` field that's somewhere in the `test::TestDescAndFn` struct initializer.
I'm not really sure how I'd then get the span of the `#[should_panic]` attribute though, as I don't think this is anywhere in the HIR after `#[test]` expansion

Reply:
All I can think of is having a pre expansion pass to get the span of the attribute, then a late pass for the actual linting. Perhaps you can match the functions together through the `testfn` field.

It's unfortunately gonna require a hacky solution no matter what

Reply:
> Another thing that's worth a shot is using `get_attrs` on the `DefId` of the function called in `testfn`. _maybe_ that'd have `should_panic`

Oh that's smart, I didn't think of that. I looked at the output of `-Zunpretty=hir-tree` and I can see a `#[should_panic]` attribute with the right span on there. That should work and is probably the least hacky way to make this work